contracts/TokenHolder.sol

```diff
@@ -330,6 +371,128 @@ contract TokenHolder is MultiSigWallet {
+     *      from CoGateway.redeem() in mosaic-contracts.
+     *
+     *      CoGateway contract address is needed for redeem functionality.
+     *      It's fetched from UtilityToken. As per requirement TokenHolder
+     *      should be able to communicate with EIP20.
+     *
+     * @param _beneficiary The address in the origin chain where the tokens
+     *                     will be released.
+     * @param _amount Redeem amount that will be transferred from tokenholder
+     *                account.
+     * @param _gasPrice Gas price that tokenholder is ready to pay to get the
+     *                  redemption process done.
+     * @param _gasLimit Gas limit that tokenholder is ready to pay.
+     * @param _redeemerNonce Nonce of the redeemer address.
+     * @param _hashLock Hash Lock provided by the facilitator.
+     * @param _nonce The nonce of an ephemeral key that was used to sign
+     *               the transaction.
+     * @param _v V of the signature.
+     * @param _r R of the signature.
+     * @param _s S of the signature.
+     *
+     * @return executionStatus_ which is bool status of coGateway.redeem.
+     */
+    function redeem(
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _gasPrice,
+        uint256 _gasLimit,
+        uint256 _redeemerNonce,
+        bytes32 _hashLock,
+        uint256 _nonce,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+    )
+        public
+        payable
+        returns (bool executionStatus_)
+    {
+        address coGateway = token.coGateway();
+
+        address ephemeralKey = verifyRedeemExecutableTransaction(
+            _amount,
+            _beneficiary,
+            _gasPrice,
+            _gasLimit,
+            _redeemerNonce,
+            coGateway,
+            _nonce,
+            _v,
+            _r,
+            _s
+        );
+
+        require(
+            _amount <= getSpendingLimit(ephemeralKey),
+            "Amount to redeem should be lte to spending limit."
+        );
+
+        token.approve(coGateway, getSpendingLimit(ephemeralKey));
+
+        bytes memory data = getCoGatewayRedeemExecutableData(
+            _amount,
+            _beneficiary,
+            _gasPrice,
+            _gasLimit,
+            _redeemerNonce,
+            _hashLock
+        );
+
+        // solium-disable-next-line security/no-call-value
+        executionStatus_ = coGateway.call.value(msg.value)(data);
+
+        token.approve(coGateway, 0);
+
+        emit RedeemInitiated(
+            _beneficiary,
+            _amount,
+            _redeemerNonce,
+            ephemeralKey,
+            executionStatus_
+        );
+    }
+
+    /**
+     * @dev Retrieves the first 4 bytes of input byte array into byte4.
+     *      Function requires:
+     *          - Input byte array's length is greater than or equal to 4.
+     */
+    function bytesToBytes4(bytes _input) public pure returns (bytes4 out_) {
```

Comment:
We already have ticket exploring library contract for TH gas optimizations. This can be taken as part of below tickets.
https://github.com/OpenSTFoundation/openst-contracts/issues/94
https://github.com/OpenSTFoundation/openst-contracts/issues/98 (Created now with more specifications)